### PR TITLE
fix(angular): item styling when control has value

### PR DESCRIPTION
**Pull request by sean-perkins**
## Pull request checklist

Please check if your PR fulfills the following requirements:
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [x] Build (`npm run build`) was run locally and any changes were pushed
- [x] Lint (`npm run lint`) has passed locally and any fixes were made for failures

## Pull request type

 

 

Please check the type of change your PR introduces:
- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe): 

## What is the current behavior?

When a control, inside an `ion-item` is displayed within a modal; after opening, closing and re-opening the modal, the `item-has-value` class is not applied to the `ion-item` that wraps the control.

Issue URL: #23809

## What is the new behavior?

- `item-has-value` is manually applied to the `ion-item` from the control value accessor, when applying other Ionic classes

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Other information

This PR does misc clean-up to deprecated `substr` and usage for add/remove classes. 

Closing #24401 in favor of this PR.

### CODE REVIEW

**Review by amandaesmith3 — APPROVED** [review]

File: angular/src/directives/control-value-accessors/value-accessor.ts [hunk 1]
```diff
@@ -110,13 +110,17 @@ export class ValueAccessor implements ControlValueAccessor, AfterViewInit, OnDes
 
 export const setIonicClasses = (element: ElementRef): void => {
   raf(() => {
-    const input = element.nativeElement as HTMLElement;
+    const input = element.nativeElement as HTMLInputElement;
+    const hasValue = typeof input.value !== 'undefined' && input.value.toString().length > 0;
```
Comment (liamdebeasi) [inline angular/src/directives/control-value-accessors/value-accessor.ts:114]: We need a null check here as the `value` prop is nullable: https://github.com/ionic-team/ionic-framework/blob/main/core/src/components/input/input.tsx#L194

Comment (liamdebeasi) [reply angular/src/directives/control-value-accessors/value-accessor.ts:114]: We should probably also write tests for when the value is `undefined` and `null` so this does not regress

**Review by liamdebeasi — CHANGES_REQUESTED** [review]

File: angular/src/directives/control-value-accessors/value-accessor.ts (hunk 1 above)
Comment (sean-perkins) [reply angular/src/directives/control-value-accessors/value-accessor.ts:114]: The internals of `ion-input` don't check for `null`: https://github.com/ionic-team/ionic-framework/blob/main/core/src/components/input/input.tsx#L327-L330

While you can assign `null` as a value, the evaluated value will always be a string (`''` for undefined/null). This aligns with how the HTML input element functions.

Do we still need to cover those cases?

Comment (liamdebeasi) [reply angular/src/directives/control-value-accessors/value-accessor.ts:114]: From what it looks like `getValue` is only used when parsing the value internally:
Usage 1: https://github.com/ionic-team/ionic-framework/blob/main/core/src/components/input/input.tsx#L236
Usage 2: https://github.com/ionic-team/ionic-framework/blob/main/core/src/components/input/input.tsx#L426
Usage 3: https://github.com/ionic-team/ionic-framework/blob/main/core/src/components/input/input.tsx#L431

None of these usages update the `value` property to be the empty string, and your code is checking the `value` prop, not calling `getValue`.

I think it's worth testing just to be sure. If we convert `null` to the empty string elsewhere such that this is not a concern, then we can proceed without the `null` check. At the moment it is not clear if that is the case.

**Review by liamdebeasi — APPROVED** [review]

**Comment by liamdebeasi** [comment]
Prior to merging, please provide the developers on https://github.com/ionic-team/ionic-framework/issues/23809 with a dev build. I'd prefer to have some developer feedback before we ship this.

**Comment by liamdebeasi** [comment]
@sean-perkins We can probably merge this now. We've given developers time to test and this new approach seems safer than the previous approach.
